### ISSUE TITLE
Silence in audio clips

### ISSUE THREAD

**Issue by WinterSoilder** [suno-ai/bark#491]
The audio clips that get generated are exactly of 14 seconds and i know it is due the batch size/array size. 
But some sentences have less words so if the clip is of 8s, there is 6s of silence.
Is there any condition to know how to crop the "silence". Or we can get dynamically generate audios with exact time the speaker speaks?

**Comment by WinterSoilder** [comment]
@gkucsko

**Comment by Tamour-Bilal** [comment]
you can use noise reduce or VAD technque for removing noise and silence between the sounds. Keep it mind, while doing postprocessing, the VAD does not work at sample rate of 24 KHZ, so u need to upsample or downsample to different frequecny. Below is the all reference postprocessing i am doing
lass BarkAudioPostProcessor:
    def __init__(self, cutoff_freq=8000, filter_order=5, use_vad=True, use_denoise=True):
        self.cutoff_freq = cutoff_freq
        self.filter_order = filter_order
        self.use_vad = use_vad
        self.use_denoise = use_denoise

    @staticmethod
    def resample_audio(audio_array, orig_sr, target_sr=16000):
        """Resample audio using polyphase filtering."""
        if orig_sr == target_sr:
            return audio_array, orig_sr
        gcd = np.gcd(orig_sr, target_sr)
        up = target_sr // gcd
        down = orig_sr // gcd
        resampled = resample_poly(audio_array, up, down)
        return resampled.astype(np.float32), target_sr

    @staticmethod
    def apply_denoising(audio_array, sample_rate):
        """Noise reduction using spectral gating (noisereduce)."""
        try:
            reduced = nr.reduce_noise(
                y=audio_array,
                sr=sample_rate,
                prop_decrease=0.9,
                stationary=False
            )
            return reduced.astype(np.float32), sample_rate
        except Exception as e:
            print(f"❌ Error in noise reduction: {e}")
            return audio_array, sample_rate

    @staticmethod
    def silero_postprocessor(audio_array, sample_rate):
        """
        Process audio using Silero VAD (remove non-speech segments).
        Silero supports 8kHz / 16kHz / 32kHz / 48kHz, but not 24kHz.
        """
        try:
            # ✅ Resample if needed
            audio_array, sr = BarkAudioPostProcessor.resample_audio(audio_array, sample_rate, 16000)

            # Convert to tensor
            if not torch.is_tensor(audio_array):
                audio_tensor = torch.from_numpy(audio_array).float()
            else:
                audio_tensor = audio_array.float()

            # Ensure mono
            if len(audio_tensor.shape) > 1:
                audio_tensor = torch.mean(audio_tensor, dim=0)

            # Run VAD
            speech_timestamps = get_speech_timestamps(
                audio_tensor,
                vad_model,
                sampling_rate=sr,
                return_seconds=False
            )

            # Extract speech
            if speech_timestamps:
                segments = [audio_tensor[t['start']:t['end']] for t in speech_timestamps]
                processed_audio = torch.cat(segments, dim=0)
            else:
                print("⚠️ Warning: No speech segments detected")
                processed_audio = audio_tensor

            return processed_audio.numpy(), sr

        except Exception as e:
            print(f"❌ Error in Silero postprocessing: {e}")
            return audio_array, sample_rate  # Return original if failed

    def process(self, audio_array, sample_rate=24000):
        """
        Full pipeline: VAD → Denoising → Cleanup.
        """
        # Step 1: Voice Activity Detection
        if self.use_vad:
            audio_array, sample_rate = self.silero_postprocessor(audio_array, sample_rate)

        # Step 2: Noise Reduction
        if self.use_denoise:
            audio_array, sample_rate = self.apply_denoising(audio_array, sample_rate)

        # Step 3: Cleanup GPU cache
        gc.collect()
        if torch.cuda.is_available():
            torch.cuda.empty_cache()

        return audio_array.astype(np.float32), sample_rate
